src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,1055 @@
+                    f"Mismatched inputs: nqubits given as {nqubits}. {class_qubit_gate} is a {qb_gate}-qubit gate."
+                )
+
+    # GST for empty circuit or operators
+    if nqubits is not None:
+        if backend is None:  # pragma: no cover
+            backend = GlobalBackend()
+
+        if nqubits == 1:
+            matrix_jk = np.zeros((4, 4))
+            for k in range(0, 4):
+                # Prepare state using "prepare_1qb_state(k)"
+                circ = prepare_states(k, nqubits)
+
+                # If class_qubit_gate is present (with/without theta)
+                if class_qubit_gate is not None:
+                    if theta is None:
+                        circ.add(getattr(gates, class_qubit_gate)(0))
+                    elif theta is not None:
+                        circ.add(getattr(gates, class_qubit_gate)(0, theta))
+
+                # Measurements
+                for j in range(0, 4):
+                    new_circ = measurement_basis(j, circ, nqubits)
+                    if noise_model is not None and backend.name != "qibolab":
+                        new_circ = noise_model.apply(new_circ)
+
+                    expectation_val = GST_execute_circuit(
+                        new_circ, k, j, nshots, nqubits
+                    )
+                    matrix_jk[j, k] = expectation_val
+
+            return matrix_jk
+
+        elif nqubits == 2:
+            matrix_jk = np.zeros((16, 16))
+            for k in range(0, 16):
+                # Prepare states using function "prepare_2qb_states(k)".
+                circ = prepare_states(k, nqubits)
+
+                # If class_qubit_gate is present (with/without theta).
+                if class_qubit_gate is not None:
+                    if theta is None:
+                        circ.add(
+                            getattr(gates, class_qubit_gate)(targ_qb[0], ctrl_qb[0])
+                        )
+                    else:
+                        circ.add(
+                            getattr(gates, class_qubit_gate)(
+                                targ_qb[0], ctrl_qb[0], theta
+                            )
+                        )
+
+                # Measurements
+                for j in range(0, 16):
+                    new_circ = measurement_basis(j, circ, nqubits)
+                    if noise_model is not None and backend.name != "qibolab":
+                        new_circ = noise_model.apply(new_circ)
+
+                    expectation_val = GST_execute_circuit(
+                        new_circ, k, j, nshots, nqubits
+                    )
+                    matrix_jk[j, k] = expectation_val
+
+            return matrix_jk
+
+        else:
+            raise ValueError(
+                f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+            )
+
+    else:
+        raise ValueError(
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+        )
+
+
+def GST_basis_operations(
```

Comment:
I don't really understand the need for this function. I mean, as I understand it, this function performs GST when the gates are in the `BasisOps` list. I assume that afterward, you want to use this basis for probabilistic error mitigation. In this case, I think it would be better to define a general method in this PR and then define this basis when implementing PEC. The GST function already does all the work. You just need to define a new function that calls `GST()` without `gate` to calculate what you call `gjk` and then for each of the gates in the gate set. After that, all you need to do is calculate the Pauli Transfer Matrix as you did at the end of `GST_basis_operations()`, fixing the gauge using the matrix `T`.

Let me know if I'm wrong and this function plays a different role.

Reply:
I have restructured this function so that now we first calculate all the basis operations and their names. Then, you just need to call `GST()`



```suggestion
def GST_basis_operations(
    nqubits=None, gjk=None, nshots=int(1e4), noise_model=None, backend=None
):
    """Runs gate set tomography for the basis operations. There are 13 basis operations for a single
        qubit and 241 basis operations for 2 qubits.

        Args:
        nshots (int, optional): Number of shots used in Gate Set Tomography.
        noise_model (:class:`qibo.noise.NoiseModel`, optional): Noise model applied
            to simulate noisy computation.
        gjk (numpy.array): Numpy array with elements Tr(Q_j rho_k) for with no gate.
            If None, call GST function with gate=None.
        backend (:class:`qibo.backends.abstract.Backend`, optional): Calculation engine.

        The 13 and 241 basis operations are based on Physical Review Research 3, 033178 (2021).
    Returns:
        numpy.array: Numpy array with elements jk equivalent to either Tr(Q_j rho_k) or Tr(Q_j O_l rho_k).
    """

    if backend is None:  # pragma: no cover
        backend = GlobalBackend()

        if nqubits != 1 and nqubits != 2:
            raise ValueError(
                f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
            )
        
        hadamard = gates.H(0).matrix(backend=backend)
        identity = gates.I(0).matrix(backend=backend)
        xgate = gates.X(0).matrix(backend=backend)
        ygate = gates.Y(0).matrix(backend=backend)
        zgate = gates.Z(0).matrix(backend=backend)
        sgate = gates.S(0).matrix(backend=backend)
        sdggate = gates.SDG(0).matrix(backend=backend)
        CNOT = gates.CNOT(0, 1).matrix(backend=backend)
        SWAP = gates.SWAP(0, 1).matrix(backend=backend)
        ry_piby4 = gates.RY(0, np.pi / 4).matrix(backend=backend)

        BasisOps_1q = [
            identity,
            xgate,
            ygate,
            zgate,
            1 / np.sqrt(2) * np.array([[1, 1j], [1j, 1]]),
            1 / np.sqrt(2) * np.array([[1, 1], [-1, 1]]),
            1 / np.sqrt(2) * np.array([[(1 + 1j), 0], [0, (1 - 1j)]]),
            1 / np.sqrt(2) * np.array([[1, -1j], [1j, -1]]),
            hadamard,
            1 / np.sqrt(2) * np.array([[0, (1 - 1j)], [(1 + 1j), 0]]),
            gates.H(0).matrix(backend=backend), #|+>
            np.array([[1, 1], [1j, -1j]]) / np.sqrt(2), #|+y>
            gates.I(0).matrix(backend=backend), #|0>
        ]

        if nqubits == 1:
            BasisOps = BasisOps_1q

            gatenames = [f"BasisOp{j}" for j in range(BasisOps)]
            gatenames[10] += "_sp_0"
            gatenames[11] += "_sp_0"
            gatenames[12] += "_sp_0"

        elif nqubits == 2:
            BasisOps_1q.extend(
                [gates.H(0).matrix(backend=backend), #|+>
                np.array([[1, 1], [1j, -1j]]) / np.sqrt(2), #|+y>
                gates.I(0).matrix(backend=backend),] #|0>
            )

            interm_ops = [
                CNOT,
                np.kron(xgate, identity) @ CNOT @ np.kron(xgate, identity),
                np.array(
                    np.block([[identity, np.zeros((2, 2))], [np.zeros((2, 2)), sgate]])
                ),
                np.array(
                    np.block(
                        [[identity, np.zeros((2, 2))], [np.zeros((2, 2)), hadamard]]
                    )
                ),
                np.kron(ry_piby4, identity) @ CNOT @ np.kron(ry_piby4, identity),
                CNOT @ np.kron(hadamard, identity),
                SWAP,
                np.array([[1, 0, 0, 0], [0, 0, 1j, 0], [0, 1j, 0, 0], [0, 0, 0, 1]]),
                SWAP @ np.kron(hadamard, identity),
            ]

            # Two qubit basis operations
            # "Easy"
            BasisOps_2q_easy = []
            for i in range(len(BasisOps_1q)):
                for j in range(len(BasisOps_1q)):
                    temp_matrix = np.kron(BasisOps_1q[i], BasisOps_1q[j])
                    BasisOps_2q_easy.append(temp_matrix)

            # "Hard"
            BasisOps_2q_hard = []
            for ii in range(0, 9):
                if ii == 6:
                    temp_matrix1 = (
                        np.kron(identity, sgate @ hadamard)
                        @ interm_ops[ii]
                        @ np.kron(identity, hadamard @ sdggate)
                    )
                    temp_matrix2 = (
                        np.kron(identity, hadamard @ sdggate)
                        @ interm_ops[ii]
                        @ np.kron(identity, sgate @ hadamard)
                    )
                    temp_matrix3 = (
                        np.kron(identity, identity)
                        @ interm_ops[ii]
                        @ np.kron(identity, identity)
                    )

                    BasisOps_2q_hard.append([temp_matrix1, temp_matrix2, temp_matrix3])

                elif ii == 7:
                    temp_matrix1 = (
                        np.kron(sgate @ hadamard, sgate @ hadamard)
                        @ interm_ops[ii]
                        @ np.kron(hadamard @ sdggate, hadamard @ sdggate)
                    )
                    temp_matrix2 = (
                        np.kron(sgate @ hadamard, hadamard @ sdggate)
                        @ interm_ops[ii]
                        @ np.kron(hadamard @ sdggate, sgate @ hadamard)
                    )
                    temp_matrix3 = (
                        np.kron(sgate @ hadamard, identity)
                        @ interm_ops[ii]
                        @ np.kron(hadamard @ sdggate, identity)
                    )
                    temp_matrix4 = (
                        np.kron(identity @ hadamard, sgate @ hadamard)
                        @ interm_ops[ii]
                        @ np.kron(identity @ sdggate, hadamard @ sdggate)
                    )
                    temp_matrix5 = (
                        np.kron(identity @ hadamard, hadamard @ sdggate)
                        @ interm_ops[ii]
                        @ np.kron(identity @ sdggate, sgate @ hadamard)
                    )
                    temp_matrix6 = (
                        np.kron(identity, identity)
                        @ interm_ops[ii]
                        @ np.kron(identity, identity)
                    )

                    BasisOps_2q_hard.append([temp_matrix1, temp_matrix2, temp_matrix3, temp_matrix4, temp_matrix5, temp_matrix6])
                else:
                    temp_matrix1 = (
                        np.kron(sgate @ hadamard, sgate @ hadamard)
                        @ interm_ops[ii]
                        @ np.kron(hadamard @ sdggate, hadamard @ sdggate)
                    )
                    temp_matrix2 = (
                        np.kron(sgate @ hadamard, hadamard @ sdggate)
                        @ interm_ops[ii]
                        @ np.kron(hadamard @ sdggate, sgate @ hadamard)
                    )
                    temp_matrix3 = (
                        np.kron(sgate @ hadamard, identity)
                        @ interm_ops[ii]
                        @ np.kron(hadamard @ sdggate, identity)
                    )

                    temp_matrix4 = (
                        np.kron(hadamard @ sdggate, sgate @ hadamard)
                        @ interm_ops[ii]
                        @ np.kron(sgate @ hadamard, hadamard @ sdggate)
                    )
                    temp_matrix5 = (
                        np.kron(hadamard @ sdggate, hadamard @ sdggate)
                        @ interm_ops[ii]
                        @ np.kron(sgate @ hadamard, sgate @ hadamard)
                    )
                    temp_matrix6 = (
                        np.kron(hadamard @ sdggate, identity)
                        @ interm_ops[ii]
                        @ np.kron(sgate @ hadamard, identity)
                    )

                    temp_matrix7 = (
                        np.kron(identity, sgate @ hadamard)
                        @ interm_ops[ii]
                        @ np.kron(identity, hadamard @ sdggate)
                    )
                    temp_matrix8 = (
                        np.kron(identity, hadamard @ sdggate)
                        @ interm_ops[ii]
                        @ np.kron(identity, sgate @ hadamard)
                    )
                    temp_matrix9 = (
                        np.kron(identity, identity)
                        @ interm_ops[ii]
                        @ np.kron(identity, identity)
                    )

                    BasisOps_2q_hard.append([temp_matrix1, temp_matrix2, temp_matrix3, temp_matrix4, temp_matrix5, temp_matrix6, temp_matrix7, temp_matrix8, temp_matrix9])

            BasisOps = BasisOps_2q_easy + BasisOps_2q_hard

            gatenames = [f"BasisOp{j}" for j in range(BasisOps)]
            gatenames[10] += "_sp_0"
            gatenames[11] += "_sp_0"
            gatenames[12] += "_sp_0"

            index = 0
            for idx_1 in range(len(BasisOps_1q)):
                for idx_2 in range(len(BasisOps_1q)):
                    # PARTIALLY REPREPARE QUBIT_0
                    if idx_1 in [10, 11, 12] and idx_2 < 10:
                        gatenames[index] += "_sp_0"
                    # PARTIALLY REPREPARE QUBIT_1
                    elif idx_2 in [10, 11, 12] and idx_1 < 10:
                        gatenames[index] += "_sp_1"
                    elif idx_1 in [10, 11, 12] and idx_2 in [10, 11, 12]:
                        gatenames[index] += "_sp_t"
                    index += 1        

        Bjk_hat = []
        for idx, gate in enumerate(BasisOps):
            gate = gates.Unitary(gate, *list(range(nqubits)))
            gate.name = gatenames[idx]
            Gk = GST(nqubits, gate, nshots, noise_model, backend=backend)
            Bjk_hat.append(Gk)
        Bjk_hat = np.array(Bjk_hat)


        # Reshape
        Bjk_hat_reshaped = np.zeros((4**(2*nqubits), len(gatenames)))
        for idx_basis_ops in range(len(gatenames)):
            temp = np.reshape(
                Bjk_hat[idx_basis_ops, :, :], [-1, 1], order="F"
            )  # 'F' so that it stacks columns.
            Bjk_hat_reshaped[:, idx_basis_ops] = temp.flatten()

        return Bjk_hat, Bjk_hat_reshaped, BasisOps
```